flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/FlinkKafkaProducer.java

```diff
@@ -226,13 +228,19 @@ public int getTransactionCoordinatorId() {
 
 	private void flushNewPartitions() {
 		LOG.info("Flushing new partitions");
+		enqueueNewPartitions().await();
+	}
+
+	private TransactionalRequestResult enqueueNewPartitions() {
 		Object transactionManager = getValue(kafkaProducer, "transactionManager");
-		Object txnRequestHandler = invoke(transactionManager, "addPartitionsToTransactionHandler");
-		invoke(transactionManager, "enqueueRequest", new Class[]{txnRequestHandler.getClass().getSuperclass()}, new Object[]{txnRequestHandler});
-		TransactionalRequestResult result = (TransactionalRequestResult) getValue(txnRequestHandler, txnRequestHandler.getClass().getSuperclass(), "result");
-		Object sender = getValue(kafkaProducer, "sender");
-		invoke(sender, "wakeup");
-		result.await();
+		synchronized (transactionManager) {
+			Object txnRequestHandler = invoke(transactionManager, "addPartitionsToTransactionHandler");
```

Comment:
Not related to your changes and can be done in a separate PR but I think we should add this check from the original sources to *"mitigate"* KAFKA-6119:
```
if (!newPartitionsInTransaction.isEmpty())
            enqueueRequest(addPartitionsToTransactionHandler());
```

Reply:
I implemented it, however in that case it would be much harder to test reliably `enqueueRequest(addPartitionsToTransactionHandler());` since in vast majority of cases this would be a dead code :/

Reply:
How are you testing now that `enqueueRequest` is called?

Reply:
It is always being called on `FlinkKafkaProducer::flush()`, so that's pretty easy :)

Reply:
Did you add the `!newPartitionsInTransaction.isEmpty()` check in the end? I couldn't find it on first glance. 

Regarding tests, if you add the check, would your current test fail? If not, I think the behaviour isn't properly tested.